Separate address hashing

nc6 opened the pull request:
Draft only at the moment for looking-at-changes purposes.

JaredCorduan commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:
```diff
@@ -99,36 +130,79 @@ import Shelley.Spec.Ledger.Crypto
+  BlockIssuer :: KeyRole RegularHash
+  -- Witness keys may have either hash type.
+  AWitness :: KeyRole AddrHash
+  RWitness :: KeyRole RegularHash
```
oh my, this is gorgeous! :old_key:

JaredCorduan reviewed (commented):
this is wonderful @nc6, I love it

nc6 committed:
Separate hashing for addresses

- We now enhance the key role with a separate parameter indicating its
hash type. This allows us to distinguish between keys which are hashed
in different ways. We have two hash types - address, and regular.
- The `keyHash` function now performs the correct hash depending on the
type of the key.
- Witnessing now needs to distinguish between the various types of keys
used. So we now have various witness sets which contain both address and
regular witnesses.

These changes caused some problems for the tests, which freely mix key
roles all over the place. So we have also refactored the tests to be
much clearer about which keys are being used and how.

The `NonTraceProperties` are even worse than the normal tests in terms
of key usage, and here we give in temporarily and use `unsafeCoerce`,
which is safe in this case since we know that the hash types align. At a
lower priority, we should rewrite these tests to use `KeySpace`.

The `Address` test is currently disabled, since it did some strange
things mixing crypto types. We will open a ticket to rationalise and
re-enable it.

Fixing the key tests seemed to cause a breakage in one of our property
tests (registeredPoolIsAdded). On closer inspection, this test was
checking invalid properties, so it's a little worrying that it was
passing before. My best hypothesis is that the confusion about keys was
preventing pools from ever being _re-registered_ in this test, which
would have allowed it to pass. In any case, this was incorrect, so this
PR also includes a fix for that test.

redxaxder commented on shelley/chain-and-ledger/executable-spec/cddl-files/shelley.cddl:
```diff
@@ -186,8 +186,9 @@ protocol_param_update =
   { ?0 => [* vkeywitness ]
-  , ?1 => [* multisig_script ]
-  , ?2 => [* bootstrap_witness ]
+  , ?1 => [* vkeywitness ]
```
Perhaps one of these could be renamed to address_vkey_witness

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:
```diff
@@ -99,36 +130,79 @@ import Shelley.Spec.Ledger.Crypto
+
+instance Crypto crypto => IsKeyRole AWitness crypto
+
+instance Crypto crypto => IsKeyRole RWitness crypto
```
There must be a way to do this

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Keys.hs:
```diff
@@ -199,32 +279,54 @@ verifySignedDSIGN (VKey vk) vd sigDSIGN =
+    Typeable (KeyRoleHashType disc),
+    Hash.HashAlgorithm (AlgorithmForHashType crypto (KeyRoleHashType disc))
+  ) =>
+  ToCBOR (KeyHash disc crypto)
```
IsKeyRole

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs:
```diff
@@ -212,10 +212,10 @@ initialFundsPseudoTxIn addr =
     pseudoTxIn h = TxIn (pseudoTxId h) 0
-    pseudoTxId :: Crypto.Hash (HASH c) a -> TxId c
+    pseudoTxId :: Crypto.Hash x a -> TxId c
     pseudoTxId = TxId . castHash
```
This looks suspicious. What happens if we have the wrong length?

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs:
```diff
@@ -757,13 +809,12 @@ witsVKeyNeeded utxo' tx@(Tx txbody _ _) _genDelegs =
         ]
-    certAuthors = Set.map asWitness . Set.fromList $ foldl (++) [] $ fmap getCertHK certificates
-    getCertHK = cwitness
+    certAuthors = foldl' (<>) ([], []) $ fmap cwitness certificates
```
Accidentally quadratic?

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/STS/Utxow.hs:
```diff
@@ -229,8 +236,11 @@ utxoWitnessed =
+          missingWitnesses =
+            WitHashes
+              (addrWitHashes needed `Set.difference` addrWitHashes witsKeyHashes)
+              (regWitHashes needed `Set.difference` regWitHashes witsKeyHashes)
```
diffWitHashes

redxaxder commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Serialization.hs:
```diff
@@ -97,6 +97,7 @@ import Data.Text (Text)
 import qualified Data.Text as Text
 import Data.Typeable
 import Network.Socket (HostAddress6)
+import Prelude
```
This is mysterious

redxaxder commented on shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs:
```diff
@@ -262,10 +272,28 @@ genAddr =
+          ( KeyRoleHashType kr
+          )
+      )
+  ) =>
```
IsKeyRole

redxaxder commented on shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Address.hs:
```diff
@@ -262,10 +272,28 @@ genAddr =
+          ( KeyRoleHashType kr
+          )
+      )
+  ) =>
```
IsKeyRole

redxaxder commented on shelley/chain-and-ledger/executable-spec/test/shelley-spec-ledger-test.cabal:
```diff
@@ -53,6 +54,7 @@ library
   build-depends:
     aeson,
     base,
+    base16-bytestring,
```
Is this used?

nc6 replied:
Turns out it's needed by the Fees module

redxaxder commented on shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/Generator/Genesis.hs:
```diff
@@ -150,8 +162,14 @@ genSeed :: Int -> Gen Seed
+          c
+          (KeyRoleHashType disc)
+      )
+  ) =>
```
IsKeyRole

redxaxder commented on shelley/chain-and-ledger/executable-spec/test/Test/Shelley/Spec/Ledger/NonTraceProperties/Generator.hs:
```diff
@@ -391,13 +392,13 @@ genDCertDeRegKey keys =
 genDCertRetirePool keys epoch = do
   key <- getAnyStakeKey keys
-  pure $ DCertPool $ RetirePool (coerceKeyRole $ hashKey key) epoch
+  pure $ DCertPool $ RetirePool (unsafeCoerce $ hashKey key) epoch
```
Perhaps `coerce` applies here.

nc6 committed:
Ormol

JaredCorduan reviewed (approved):
:+1:

nc6 commented:
bors merge
